### PR TITLE
feat: add co-build integration in code base

### PR DESCRIPTION
**Pull request by ashuralyk**
# Description
co-build integration for Spore contract.

`Address` in Spore action definition means the intense of users if to check corresponding address, like `from` and `to`, since each Spore transaction has its `from` or `to` address, we require co-build related contracts to check their validation of addresses.

I've deleted and adjusted some duplicated and misunderstanding test cases for simplicity.

`tests` has been refactored.

# Tasks list
- [x] integrate co-build in Spore contract code base
- [x] adjust test cases for adapting co-build check

### CODE REVIEW

File: lib/utils/src/lib.rs [hunk 1]
```diff
@@ -3,37 +3,53 @@
...
-pub fn verify_type_id(index: usize, source: Source) -> bool {
+pub fn load_type_args(index: usize, source: Source) -> Bytes {
+    load_cell_type(index, source)
+        .unwrap_or(None)
+        .unwrap_or_default()
+        .args()
+        .raw_data()
+}
+
+pub fn verify_type_id(index: usize, source: Source) -> Option<[u8; 32]> {
```
Comment (Flouse) [inline lib/utils/src/lib.rs:35]:
```suggestion
pub fn verify_type_id(output_index: usize, source: Source) -> Option<[u8; 32]> {
```

File: lib/utils/src/lib.rs [hunk 2]
```diff
@@ -3,37 +3,53 @@
...
+    }
+    if type_id_args.as_ref()[..32] == expected_id {
+        return Some(expected_id);
     }
-    type_id_args.as_ref()[..32] == expected_id
+
+    None
 }
 
 pub fn calc_type_id(prevout_hash: &[u8], output_index: usize) -> [u8; 32] {
```
Comment (Flouse) [inline lib/utils/src/lib.rs:55]:
```suggestion
pub fn calc_type_id(tx_first_input: &[u8], output_index: usize) -> [u8; 32] {
```

File: lib/utils/src/lib.rs [hunk 3]
```diff
@@ -3,37 +3,53 @@
...
+        Err(_) => return None,
     };
 
     let expected_id = calc_type_id(first_input.as_slice(), index);
-    let type_id_args = load_cell_type(index, source)
-        .unwrap_or(None)
-        .unwrap_or_default()
-        .args()
-        .raw_data();
+    let type_id_args = load_type_args(index, source);
```
Comment (Flouse) [inline lib/utils/src/lib.rs:42]:
```suggestion
    let type_id_args = load_type_args(index, Source::Output);
```

File: lib/utils/src/lib.rs (hunk 2 above)
Comment (Flouse) [reply lib/utils/src/lib.rs:55]: prevout_hash -> tx_first_input

File: contracts/spore/src/entry.rs [hunk 4]
```diff
@@ -178,6 +204,16 @@ fn process_transfer() -> Result<(), Error> {
...
 
+    // check co-build action @lyk
+    let action::SporeActionUnion::Transfer(transfer) = extract_spore_action()?.to_enum() else {
+        return Err(Error::SporeActionMismatch);
+    };
+    if transfer.nft_id().as_slice() != load_type_args(0, GroupInput).as_ref() {
+        return Err(Error::SporeActionFieldMismatch);
+    }
+    check_spore_address(GroupInput, transfer.to())?;
+    check_spore_address(GroupOutput, transfer.from())?;
```
Comment (Flouse) [inline contracts/spore/src/entry.rs:215]:
```suggestion
    check_spore_address(GroupInput, transfer.from())?;
    check_spore_address(GroupOutput, transfer.to())?;
```

File: lib/utils/src/lib.rs [hunk 5]
```diff
@@ -126,3 +143,38 @@ pub fn calc_capacity_sum(lock_hash: &[u8; 32], source: Source) -> u128 {
...
+            }
+            action = Some(value);
+        }
+    }
+    let Some(action) = action else {
+        return Err(Error::InvliadCoBuildMessage);
+    };
+    let spore_action = action::SporeAction::from_slice(&action.data().raw_data())
+        .map_err(|_| Error::InvliadCoBuildMessage)?;
+    Ok(spore_action)
```
Comment (quake) [inline lib/utils/src/lib.rs:179]: suggest to use iter
```suggestion
    let mut iter = message.actions().into_iter().filter(|value| value.script_hash().as_slice() == script_hash.as_slice());
    match (iter.next(), iter.next()) {
        (Some(action), None) => action::SporeAction::from_slice(&action.data().raw_data()).map_err(|_| Error::InvliadCoBuildMessage),
        _ => Err(Error::InvliadCoBuildMessage)
    }
```

File: tests/src/utils/co_build.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    SporeActionUnion::Mint(mint)
+}
+
+pub fn build_transfer_action(context: &mut Context, nft_id: [u8; 32]) -> SporeActionUnion {
+    let script = internal::build_always_success_script(context);
+    let address = script_to_address(script);
+    let transfer = Transfer::new_builder()
+        .nft_id(h256_to_byte32(nft_id))
+        .from(address.clone())
+        .to(address)
```
Comment (quake) [inline tests/src/utils/co_build.rs:101]: suggest to use different addresses for `from/to` in the unit test

File: tests/src/tests/cluster.rs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+use crate::Loader;
+use crate::MAX_CYCLES;
+
+#[test]
+fn test_simple_cluster_mint() {
+    let mut context = Context::default();
+
+    let cluster = build_serialized_cluster_data("Spore Cluster", "Test Cluster");
+
+    // always success lock
```
Comment (quake) [inline tests/src/tests/cluster.rs:18]: is this comment wrong? it's not a lock script.

**Review by quake — CHANGES_REQUESTED** [review]

File: lib/utils/src/lib.rs [hunk 8]
```diff
@@ -126,3 +146,32 @@ pub fn calc_capacity_sum(lock_hash: &[u8; 32], source: Source) -> u128 {
...
+    let script_hash = load_script_hash()?;
+
+    let mut iter = message
+        .actions()
+        .into_iter()
+        .filter(|value| value.script_hash().as_slice() == script_hash.as_slice());
+    match (iter.next(), iter.next()) {
+        (Some(action), None) => action::SporeAction::from_slice(&action.data().raw_data())
+            .map_err(|_| Error::InvliadCoBuildMessage),
+        _ => Err(Error::InvliadCoBuildMessage),
```
Comment (Flouse) [inline lib/utils/src/lib.rs:175]: Error::SporeActionDuplicated is defined but not used.
Is it supposed to be used here?

File: lib/types/schemas/action.mol [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    nft_id: Byte32,
+    from: Address,
+    to: Address,
+}
+
+table Burn {
+    nft_id: Byte32,
+    from: Address,
+}
+
```
Comment (Flouse) [inline lib/types/schemas/action.mol:32]: ### nft_id -> spore_id

see RFC:
https://github.com/sporeprotocol/spore-contract/blob/6921c150b4004e8350a8d69110a3095398279a60/RFC.md#L30

**Review by Flouse — CHANGES_REQUESTED** [review]

**Comment by Flouse** [comment]
Is this code useless?

https://github.com/sporeprotocol/spore-contract/blob/6921c150b4004e8350a8d69110a3095398279a60/contracts/cluster/src/entry.rs#L28-L60

Never mind, maybe it could be deleted later.

File: lib/utils/src/lib.rs (hunk 8 above)
Comment (ashuralyk) [reply lib/utils/src/lib.rs:175]: oh yes, after merging jjy's suggestion, it's changed

File: lib/types/schemas/action.mol (hunk 9 above)
Comment (ashuralyk) [reply lib/types/schemas/action.mol:32]: accepted

**Review by quake — APPROVED** [review]

File: contracts/cluster_agent/src/entry.rs [hunk 10]
```diff
@@ -33,17 +38,17 @@ fn process_creation(_index: usize) -> Result<(), Error> {
...
     }
 
     // Condition 1: Check if cluster proxy exist in Inputs & Outputs
-    return if find_posityion_by_type_hash(proxy_type_hash.as_slice(), Input).is_some()
-        && find_posityion_by_type_hash(proxy_type_hash.as_slice(), Output).is_some()
-    {
-        Ok(())
-    } else {
+    let proxy_cell_in_input =
+        find_posityion_by_type_hash(proxy_type_hash.as_slice(), Input).is_some();
```
Comment (xxuejie) [inline contracts/cluster_agent/src/entry.rs:47]: find_position_by_type_hash

**Review by xxuejie — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

File: tests/src/tests/cluster.rs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    let tx = context.complete_tx(tx);
+
+    context
+        .verify_tx(&tx, MAX_CYCLES)
+        .expect("test simple spore mint");
+}
+
+#[test]
+fn test_simple_spore_mint_with_cluster() {
+    let cluster_id = H256::from_trimmed_str("12345678".clone())
```
Comment (Flouse) [inline tests/src/tests/cluster.rs:43]: Why is cluster_id defined as `from_trimmed_str("12345678"`?

### ref
CLUSTER_ID = `hash(this_transaction.inputs[0]) | Grouputput_index_of_this_cell`

**Comment by ashuralyk** [comment]
@ShookLyngs rerun CI again once `spore-sdk` enabled its co-build feature

**Review by ShookLyngs — CHANGES_REQUESTED** [review]
About the `action.mol`, I have found a few potential issues, please check the following:

1. The corresponding "melt" actions for ClusterProxy and ClusterAgent are not defined. ClusterProxy and ClusterAgent are both designed to be meltable, so maybe they should be added.
2. The `code_hash` field is not defined in both the `ProxyCreate` and the `AgentCreate` actions, but the field is defined in the `Mint` and the `ClusterCreate` actions. 
3. The `proxy_id` field is not defined in the `AgentCreate` action, but when creating a ClusterAgent, a ClusterProxy must be referenced in the transaction (via payment or cell reference).

**Comment by ShookLyngs** [comment]
I have some thoughts regarding the naming of the `SporeAction` (just some personal suggestions):
- Inconsistent names such as `Mint` and `ClusterCreate` can be renamed to align with each other.
- Personally I feel like the naming of `ClusterCreate` looks off, perhaps alternatives like `CreateCluster` or `ClusterCreation` would be more appropriate.

**Comment by ashuralyk** [comment]
> The code_hash field is not defined in both the ProxyCreate and the AgentCreate actions, but the field is defined in the Mint and the ClusterCreate actions

1. `data_hash` in proxy is not necessary, because the data itself represents `cluster_id`, which is included in molecule definition
2. `data_hash` in agent is `proxy_type_hash` that is used to locate proxy in celldep, I can add `proxy_id` to check instead

@ShookLyngs

**Comment by ShookLyngs** [comment]
> 1. `data_hash` in proxy is not necessary, because the data itself represents `cluster_id`, which is included in molecule definition
> 2. `data_hash` in agent is `proxy_type_hash` that is used to locate proxy in celldep, I can add `proxy_id` to check instead

Have I misunderstood why the data_hash is required in a creation action?
I thought it was for validating the code hashes. Could you please explain the design behind it? Thank you.

> in both proxy and agent contracts, there's no function to check validation of action `melt`, so I guess there's no need to add melt action for them

Do you mean the co-build feature should be disabled when melting ClusterProxy and ClusterAgent? 
I think I need more instructions about when it should be enabled/disabled.

**Comment by ashuralyk** [comment]
> Have I misunderstood why the data_hash is required in a creation action?

I guess no, because current co-build design is rely on us, we have responsibility to design it to match our need, so the molecule I designed is to match my need that collected from contracts implementation

I think if you have your sense of design from sdk, feel free to post it out, and we can discuss

**Comment by ShookLyngs** [comment]
> I guess no, because current co-build design is rely on us, we have responsibility to design it to match our need, so the molecule I designed is to match my need that collected from contracts implementation

In my point of view, co-build is a feature to enable typed message for CKB wallets. And to acheive this, each type of transaction related to Spore, e.g. creation, transfer, melting, should have a coressponding action defined in the molecule. Otherwise, what would happen when user is attmpting to melt a ClusterProxy? Would the wallet simply ask the user to sign for a hash (instead of a typed message)?

From my perspective, co-build is a feature that enables typed messages in ckb wallets. To achieve this, each type of Spore-related transaction, such as creation, transfer, and melting, should have a corresponding action defined in the molecule. Otherwise, I'm not sure what would happen if a user tries to melt a ClusterProxy, would the wallet simply request the user to sign a hash instead of a typed message?

Need more info on this, thx.

**Comment by ashuralyk** [comment]
> To achieve this, each type of Spore-related transaction, such as creation, transfer, and melting, should have a corresponding action defined in the molecule

I got your concerns, I'll add melt action for proxy and agent only with `from` address as parameter

**Comment by Flouse** [comment]
Cluster Proxy and Cluster Agent is planned in [Stage 2](https://dao.ckb.community/thread/vot-spore-protocol-mainnet-launch-sponsorship-proposal-48305).

Are you going to test them right now? or just focus on Stage 1 before 2024-01-20?

**Comment by ashuralyk** [comment]
> Cluster Proxy and Cluster Agent is planned in [Stage 2](https://dao.ckb.community/thread/vot-spore-protocol-mainnet-launch-sponsorship-proposal-48305).
> 
> Are you going to test them right now? or just focus on Stage 1 before 2024-01-20?

test process is the same and won't take more time, and it's better to test them all, thus we don't have to pay twice for re-deploying Cluster and Spore contracts again

**Comment by ashuralyk** [comment]
@ShookLyngs proxy/agent actions are prepared, please check it out

**Comment by ShookLyngs** [comment]
Encountered some failed cases during my tests using this branch of code: https://github.com/sporeprotocol/spore-sdk/pull/68.
On the other hand, Spore/Cluster related tests have passed.

Error 11:
- Transfer ClusterProxy
- Melt ClusterProxy
- Create ClusterAgent with ClsuterProxy (via cell reference)
- Melt ClusterAgent

I've only added detailed cobuild checks to the "Transfer ClusterProxy" test case (see here: https://github.com/sporeprotocol/spore-sdk/pull/68/files#diff-0d580b1ad4f9eff4775817d276fc19a0c95a2e00cc6263f94bc8c394fe44b411).

**Pull request merged**